From 234d3c5e7e6d77cf107902d915fea9cb2d6c4554 Mon Sep 17 00:00:00 2001
From: Matthias FRANCK <matthias.franck_ext@softathome.com>
Date: Thu, 22 Dec 2022 14:23:48 +0100
Subject: [PATCH] Update makefile

---
 src/makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/makefile b/src/makefile
index 16f9b5c..b328bb3 100644
--- a/src/makefile
+++ b/src/makefile
@@ -15,7 +15,7 @@ TARGET_A = $(OBJDIR)/$(COMPONENT).a
 SRCDIR = $(shell pwd)
 INCDIR_PUB = $(SRCDIR)/../include
 INCDIR_PRIV = $(SRCDIR)/../include_priv
-INCDIRS = $(INCDIR_PUB) $(INCDIR_PRIV) $(if $(STAGINGDIR), $(STAGINGDIR)/include)
+INCDIRS = $(INCDIR_PUB) $(INCDIR_PRIV) $(if $(STAGINGDIR), $(STAGINGDIR)/include) $(if $(STAGINGDIR), $(STAGINGDIR)/opt/prplos/usr/include)
 LIBDIR = $(if $(STAGINGDIR), -L$(STAGINGDIR)/lib -L$(STAGINGDIR)/usr/lib)
 
 # files
-- 
GitLab
